### PR TITLE
enabling `export_and_compile` for `QEFFAutoModelForCausalLM`

### PR DESCRIPTION
**Pull request by ochougul**
* Introduced `Runtime`(`CPU_PT`, `CPU_ORT`, `AI_100`)
* enabled export, compile and export_and_compile for `QEFFAutoModelForCausalLM`

### CODE REVIEW

File: QEfficient/src/common.py
Comment (vbaddi) [inline QEfficient/src/common.py:86]: Do we want to pass `**kwargs` to the `AutoModelForCausalLM `__class__, might lead to hf_token issue in the hugginface.

File: QEfficient/utils/_utils.py
Comment (vbaddi) [inline QEfficient/utils/_utils.py:122]: okay to keep default padding to "`right`"?

File: QEfficient/utils/_utils.py
Comment (vbaddi) [inline QEfficient/utils/_utils.py:128]: Need `*.txt`, to get tokenizers from models like `GPTJ `which are not in json formats.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py
Comment (vbaddi) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:388]: lets discuss this `runtime `inclusion once.

File: QEfficient/utils/_utils.py
Comment (ochougul) [reply QEfficient/utils/_utils.py:122]: will update, these changes came in, as I was reviving this old branch, and main was too ahead of this.

File: QEfficient/utils/_utils.py
Comment (ochougul) [reply QEfficient/utils/_utils.py:128]: will update. Thanks.

File: QEfficient/src/common.py
Comment (ochougul) [reply QEfficient/src/common.py:86]: Good catch. Fixed

File: QEfficient/src/_transformers/runtime_args.py
Comment (irajagop) [inline QEfficient/src/_transformers/runtime_args.py:16]: Do we need a new dataclass for runtime args? A simple dict won't work?

Comment (ochougul) [reply QEfficient/src/_transformers/runtime_args.py:16]: 1. Dataclasses automatically does type restriction of attributes
2. We are doing some check on them in __post_init__ which we will  have to write everywhere they are being initialized if we use dicts
3. Also, I am thinking of freezing this dataclasses so that, user can't assign arbitrary values to them
4. So, because of 2 and 3 we are always sure that once runtime-args are populated we can always allowed to execute in respective runtime, making it robust from failures.

I understand that dicts are more efficient for initialization and access, but I feel the gain is negligible as compared to the above advantages we are getting by using dataclass

ref: 
https://stackoverflow.com/questions/55248416/python-3-which-one-is-faster-for-accessing-data-dataclasses-or-dictionaries
https://stackoverflow.com/questions/74117873/data-classes-vs-dictionaries

File: QEfficient/src/_transformers/auto.py
Comment (irajagop) [inline QEfficient/src/_transformers/auto.py:112]: Why do we need to have only one runtime at a time? We could have something like `execute_pytorch`, `execute_onnxrt` and `execute_aic`. We could still have `self.runtime` set to something as default, that will direct which `execute_` method to be called.

Comment (irajagop) [reply QEfficient/src/_transformers/auto.py:112]: Also, we could manage the args internally no? For example, after export, we could directly update the `self.onnx_model_path` instead of needing to set it from outside.

File: QEfficient/src/_transformers/runtime_args.py
Comment (irajagop) [reply QEfficient/src/_transformers/runtime_args.py:16]: As suggested above, if we use `execute_` method with correct args or kwargs, that will work as validation. We're going to use it at only one place, for execution. `dataclass` makes more sense when we're re-using in multiple places, serializing or sending over network etc.

File: QEfficient/src/_transformers/auto.py
Comment (ochougul) [reply QEfficient/src/_transformers/auto.py:112]: If we expose `self.onnx_model_path` as model variable we allow users to set it any value they want, we can pass `forzen=True` in `dataclasses` to not allow such usage.

File: README.md
Comment (irajagop) [inline README.md:235]: typo

Comment (ochougul) [reply README.md]: updated.

File: QEfficient/generation/text_generation_inference.py
Comment (irajagop) [inline QEfficient/generation/text_generation_inference.py:143]: `Optional[int]` should be there for generation_len.

**Review by irajagop — APPROVED** [review]

File: tests/transformers/models/test_causal_lm_models.py
Comment (vbaddi) [inline tests/transformers/models/test_causal_lm_models.py:18]: Can you remove the `phi3` model card from the `config.json`, to test the CI pipeline? todo @quic-mamta for phi3 modeling debug.

Comment (vbaddi) [reply tests/transformers/models/test_causal_lm_models.py:18]: this is taken care in #63. Thanks

**Review by vbaddi — APPROVED** [review]
LGTM, Once CI Passes, pls merge. Thanks

**Pull request merged**
